File: src/main/java/com/architecture_map/belarus/entity/Architect.java
package com.architecture_map.belarus.entity;

//import com.architecture_map.belarus.entity.image.ArchitectImage;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.ToString;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.OneToOne;
import java.util.HashSet;
import java.util.Set;

@Entity
@Getter
@Setter(AccessLevel.PUBLIC)
@RequiredArgsConstructor
@EqualsAndHashCode
@ToString
public class Architect {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    private String name;
    private String middleName;
    private String surname;
    private String yearsOfLife;

    @Column(name="short_work_description", columnDefinition="TEXT")
    private String shortWorkDescription;

    private String biographicalArticle;

//    @JsonIgnoreProperties("architect")
//    @OneToOne(mappedBy = "architect")
//    private ArchitectImage image;

    @JsonIgnoreProperties("architects")
    @ManyToMany(cascade = {CascadeType.ALL})
    @JoinTable(
            name = "construction_architect",
            joinColumns = { @JoinColumn(name = "architect_id")},
            inverseJoinColumns = {@JoinColumn(name = "construction_id")}
    )
    private Set<Construction> constructions = new HashSet<>();
}

File: src/main/kotlin/by/architecturemap/belarus/entity/Architect.kt
package by.architecturemap.belarus.entity

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import jakarta.persistence.CascadeType
import jakarta.persistence.Entity
import jakarta.persistence.JoinColumn
import jakarta.persistence.JoinTable
import jakarta.persistence.ManyToMany
import jakarta.persistence.OneToOne
import jakarta.validation.constraints.NotBlank

@Entity
data class Architect(

    @NotBlank(message = "Name may not be blank")
    var name: String,

    var middleName: String? = null,

    @NotBlank(message = "Surname may not be blank")
    var surname: String,

    /**
     * There are cases when there is no the exact date the life of architect,
     * but only an approximate time, for example: "end of the 17th century - 1756", so field has a string type
     */
    @NotBlank(message = "Years of life may not be blank")
    var yearsOfLife: String,

    @NotBlank(message = "Short work description may not be blank")
    var shortWorkDescription: String? = null,

    var biographicalArticle: String? = null,

    @OneToOne
    @JoinColumn(name = "portrait_image_id", referencedColumnName = "id")
    var portraitImage: Image? = null,

    @JsonIgnoreProperties("architects")
    @ManyToMany(cascade = [CascadeType.ALL])
    @JoinTable(
        name = "construction_architect",
        joinColumns = [JoinColumn(name = "architect_id")],
        inverseJoinColumns = [JoinColumn(name = "construction_id")]
    )
    var construction: List<Construction>? = listOf()

) : BaseEntity()
